diff --git a//.gitignore b//.gitignore
index b3e835807..8dc290e36 100644
--- a//.gitignore
+++ b///.gitignore
@@ -2,3 +2,4 @@
 out/
 
 device.bazelrc
+EXTRA/
diff --git a//private/devices/google/common/tools/build_dist.sh b//private/devices/google/common/tools/build_dist.sh
index 2e77e7ca8..4e5cf5ada 100755
--- a//private/devices/google/common/tools/build_dist.sh
+++ b///private/devices/google/common/tools/build_dist.sh
@@ -41,5 +41,5 @@ trap '{ rm -f -- "$sign_file"; }' EXIT
 prebuilts/clang/host/linux-x86/clang-r487747c/bin/clang aosp/scripts/sign-file.c -lssl -lcrypto -o ${sign_file}
 find out/${DEVICE}/dist -type f -name "*.ko" \
   -exec ${sign_file} sha256 \
-  $(find out/ -type f -name "signing_key.pem") \
-  $(find out/ -type f -name "signing_key.x509") {} \;
\ No newline at end of file
+  $(find out/ -type f -name "signing_key.pem" | head -n1) \
+  $(find out/ -type f -name "signing_key.x509" | head -n1) {} \;
\ No newline at end of file


