diff --git a//.gitignore b//.gitignore
index b3e835807..8dc290e36 100644
--- a//.gitignore
+++ b///.gitignore
@@ -2,3 +2,4 @@
 out/
 
 device.bazelrc
+EXTRA/
diff --git a//private/devices/google/common/tools/build_dist.sh b//private/devices/google/common/tools/build_dist.sh
index 2e77e7ca8..4e5cf5ada 100755
--- a//private/devices/google/common/tools/build_dist.sh
+++ b///private/devices/google/common/tools/build_dist.sh
@@ -41,5 +41,5 @@ trap '{ rm -f -- "$sign_file"; }' EXIT
 prebuilts/clang/host/linux-x86/clang-r487747c/bin/clang aosp/scripts/sign-file.c -lssl -lcrypto -o ${sign_file}
 find out/${DEVICE}/dist -type f -name "*.ko" \
   -exec ${sign_file} sha256 \
-  $(find out/ -type f -name "signing_key.pem") \
-  $(find out/ -type f -name "signing_key.x509") {} \;
\ No newline at end of file
+  $(find out/ -type f -name "signing_key.pem" | head -n1) \
+  $(find out/ -type f -name "signing_key.x509" | head -n1) {} \;
\ No newline at end of file


diff --git a//.gitignore b//.gitignore
index c9c9d6877..a0e953a7b 100644
--- a//.gitignore
+++ b///.gitignore
@@ -3,3 +3,4 @@ out/
 
 device.bazelrc
 
+EXTRA/
diff --git a//private/devices/google/common/tools/build_dist.sh b//private/devices/google/common/tools/build_dist.sh
index 01211c5d8..beb996111 100755
--- a//private/devices/google/common/tools/build_dist.sh
+++ b///private/devices/google/common/tools/build_dist.sh
@@ -27,5 +27,5 @@ trap '{ rm -f -- "$sign_file"; }' EXIT
 prebuilts/clang/host/linux-x86/clang-r510928/bin/clang aosp/scripts/sign-file.c -lssl -lcrypto -o ${sign_file}
 find out/${DEVICE}/dist -type f -name "*.ko" \
   -exec ${sign_file} sha256 \
-  $(find out/ -type f -name "signing_key.pem") \
-  $(find out/ -type f -name "signing_key.x509") {} \;
\ No newline at end of file
+  $(find out/ -type f -name "signing_key.pem" | head -n1) \
+  $(find out/ -type f -name "signing_key.x509" | head -n1) {} \;
\ No newline at end of file


diff --git a/aosp/arch/arm64/kvm/arm.c b/aosp/arch/arm64/kvm/arm.c
index 772a2847a8a5..5cd1f10a8ae2 100644
--- a/aosp/arch/arm64/kvm/arm.c
+++ b/aosp/aosp/arch/arm64/kvm/arm.c
@@ -22,6 +22,7 @@
 #include <linux/sched/stat.h>
 #include <linux/shrinker.h>
 #include <linux/psci.h>
+#include <linux/security.h>
 #include <trace/events/kvm.h>
 
 #define CREATE_TRACE_POINTS
@@ -2181,9 +2182,13 @@ static int __init init_subsystems(void)
 
 	kvm_register_perf_callbacks(NULL);
 
-	err = hyp_trace_init_tracefs();
-	if (err)
-		kvm_err("Failed to initialize Hyp tracing\n");
+	if (security_locked_down(LOCKDOWN_TRACEFS)) {
+		kvm_err("Hyp tracing disabled due to lockdown\n");
+	} else {
+		err = hyp_trace_init_tracefs();
+		if (err)
+			kvm_err("Failed to initialize Hyp tracing\n");
+	}
 out:
 	if (err)
 		hyp_cpu_pm_exit();
diff --git a/aosp/drivers/usb/gadget/udc/core.c b/aosp/drivers/usb/gadget/udc/core.c
index 22f0b24b6b36..4e54875d7aa9 100644
--- a/aosp/drivers/usb/gadget/udc/core.c
+++ b/aosp/aosp/drivers/usb/gadget/udc/core.c
@@ -714,12 +714,6 @@ static int usb_gadget_connect_locked(struct usb_gadget *gadget)
 		goto out;
 	}
 
-	if (deny_new_usb) {
-		dev_err(&gadget->dev, "blocked USB gadget connection\n");
-		ret = -EPERM;
-		goto out;
-	}
-
 	if (gadget->deactivated || !gadget->udc->allow_connect || !gadget->udc->started) {
 		/*
 		 * If the gadget isn't usable (because it is deactivated,


