diff --git a//.gitignore b//.gitignore
index 157a0367f..f41726a7a 100644
--- a//.gitignore
+++ b///.gitignore
@@ -1,3 +1,4 @@
 /bazel-*
 out/
 
+EXTRA/
diff --git a//aosp b//aosp
index 8fd157b97..721cc36ce 160000
--- a//aosp
+++ b///aosp
@@ -1 +1 @@
-Subproject commit 8fd157b971caff4af1fc982ba12df440e270d28b
+Subproject commit 721cc36ce2abd43bd457743165beb251ce97672d
diff --git a//common/ack b//common/ack
new file mode 160000
index 000000000..d3e89d1ef
--- /dev/null
+++ b///common/ack
@@ -0,0 +1 @@
+Subproject commit d3e89d1efe508483dd659e5a4d63e82a3c41c036


