diff --git a//.gitignore b//.gitignore
index 157a0367f..f41726a7a 100644
--- a//.gitignore
+++ b///.gitignore
@@ -1,3 +1,4 @@
 /bazel-*
 out/
 
+EXTRA/
diff --git a//aosp b//aosp
index 6a4b09d43..25cb517c9 160000
--- a//aosp
+++ b///aosp
@@ -1 +1 @@
-Subproject commit 6a4b09d432cbc67e439babddc8a02d7e02ca9f87
+Subproject commit 25cb517c9df8683f8171c0665670784bad4e8977
diff --git a//common/ack b//common/ack
new file mode 160000
index 000000000..d3e89d1ef
--- /dev/null
+++ b///common/ack
@@ -0,0 +1 @@
+Subproject commit d3e89d1efe508483dd659e5a4d63e82a3c41c036


