diff --git a//.gitignore b//.gitignore
index 157a0367f..f41726a7a 100644
--- a//.gitignore
+++ b///.gitignore
@@ -1,3 +1,4 @@
 /bazel-*
 out/
 
+EXTRA/
diff --git a//aosp b//aosp
index e78e90da6..b9e7fb448 160000
--- a//aosp
+++ b///aosp
@@ -1 +1 @@
-Subproject commit e78e90da69f48c01fb7ecec247497daaef7fe59e
+Subproject commit b9e7fb448368c7e220d13932b80bcf71daf2fa34
diff --git a//common/ack b//common/ack
new file mode 160000
index 000000000..d3e89d1ef
--- /dev/null
+++ b///common/ack
@@ -0,0 +1 @@
+Subproject commit d3e89d1efe508483dd659e5a4d63e82a3c41c036


