diff --git a//.gitignore b//.gitignore
index 157a0367f..f41726a7a 100644
--- a//.gitignore
+++ b///.gitignore
@@ -1,3 +1,4 @@
 /bazel-*
 out/
 
+EXTRA/
diff --git a//common/ack b//common/ack
index 009a4f15a..524c23aac 160000
--- a//common/ack
+++ b///common/ack
@@ -1 +1 @@
-Subproject commit 009a4f15a8290751157229b3024eaf1ba7f10e8b
+Subproject commit 524c23aac06235097f4205571741eb0e8e5b43b0


diff --git a/aosp/BUILD.bazel b/aosp/BUILD.bazel
index 9240837a9a8e..d5c90df09987 100644
--- a/aosp/BUILD.bazel
+++ b/aosp/aosp/BUILD.bazel
@@ -12,7 +12,6 @@ load(
     "kernel_abi",
     "kernel_build",
     "kernel_images",
-    "kernel_module_group",
     "kernel_modules_install",
     "merged_kernel_uapi_headers",
 )
@@ -131,7 +130,6 @@ filegroup(
         "android/abi_gki_aarch64_unisoc",
         "android/abi_gki_aarch64_virtual_device",
         "android/abi_gki_aarch64_vivo",
-        "android/abi_gki_aarch64_wonder",
         "android/abi_gki_aarch64_xiaomi",
         "android/abi_gki_aarch64_zebra",
     ],
@@ -316,51 +314,6 @@ copy_to_dist_dir(
     log = "info",
 )
 
-kernel_build(
-    name = "wonder_aarch64",
-    outs = [],
-    base_kernel = ":kernel_aarch64",
-    build_config = "build.config.wonder",
-    collect_unstripped_modules = True,
-    kmi_symbol_list = "android/abi_gki_aarch64_wonder",
-    make_goals = ["modules"],
-    module_outs = [
-        "net/mac80211/mac80211.ko",
-        "net/wireless/cfg80211.ko",
-    ],
-    strip_modules = True,
-)
-
-kernel_module_group(
-    name = "wonder_ddk_modules",
-    srcs = ["//common-modules/wonder"],
-)
-
-kernel_abi(
-    name = "wonder_aarch64_abi",
-    kernel_build = ":wonder_aarch64",
-    kernel_modules = [":wonder_ddk_modules"],
-)
-
-kernel_modules_install(
-    name = "wonder_aarch64_modules_install",
-    kernel_build = ":wonder_aarch64",
-    kernel_modules = [":wonder_ddk_modules"],
-)
-
-copy_to_dist_dir(
-    name = "wonder_aarch64_dist",
-    data = [
-        ":kernel_aarch64",
-        ":kernel_aarch64_additional_artifacts",
-        ":wonder_aarch64",
-        ":wonder_aarch64_modules_install",
-    ],
-    dist_dir = "out/wonder_aarch64/dist",
-    flat = True,
-    log = "info",
-)
-
 _DB845C_MODULE_OUTS = [
     # keep sorted
     "crypto/michael_mic.ko",
diff --git a/aosp/android/abi_gki_aarch64.stg b/aosp/android/abi_gki_aarch64.stg
index 736ddbbb615a..bddd2965dc41 100644
--- a/aosp/android/abi_gki_aarch64.stg
+++ b/aosp/aosp/android/abi_gki_aarch64.stg
@@ -326725,14 +326725,6 @@ function {
   parameter_id: 0x1847a854
   parameter_id: 0x4585663f
 }
-function {
-  id: 0x9a878ec0
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x4585663f
-  parameter_id: 0xf1a6dfed
-  parameter_id: 0x92233392
-}
 function {
   id: 0x9a8b3be7
   return_type_id: 0x6720d32f
@@ -328029,12 +328021,6 @@ function {
   parameter_id: 0x18bd6530
   parameter_id: 0x27162aac
 }
-function {
-  id: 0x9b3403c3
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x26ee682a
-}
 function {
   id: 0x9b342f25
   return_type_id: 0x6720d32f
@@ -328889,18 +328875,6 @@ function {
   parameter_id: 0x13580d6c
   parameter_id: 0xfc0e1dbd
 }
-function {
-  id: 0x9b65d49e
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x36f98e5c
-  parameter_id: 0x4585663f
-  parameter_id: 0x33756485
-  parameter_id: 0x92233392
-  parameter_id: 0x33756485
-  parameter_id: 0x33756485
-  parameter_id: 0x6720d32f
-}
 function {
   id: 0x9b65de16
   return_type_id: 0x6720d32f
@@ -330274,13 +330248,6 @@ function {
   return_type_id: 0x6720d32f
   parameter_id: 0x18bd6530
 }
-function {
-  id: 0x9bb0e048
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x0665e6b6
-  parameter_id: 0x1b36c7a2
-}
 function {
   id: 0x9bb12429
   return_type_id: 0x6720d32f
@@ -331435,14 +331402,6 @@ function {
   parameter_id: 0x18ea6ae3
   parameter_id: 0x18bd6530
 }
-function {
-  id: 0x9bd95971
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x1b36c7a2
-  parameter_id: 0x6720d32f
-  parameter_id: 0xfc0e1dbd
-}
 function {
   id: 0x9bd9619b
   return_type_id: 0x6720d32f
@@ -332007,14 +331966,6 @@ function {
   parameter_id: 0x18bd6530
   parameter_id: 0x1013df15
 }
-function {
-  id: 0x9bf05906
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x1582ab06
-  parameter_id: 0x2170d06d
-  parameter_id: 0x6d7f5ff6
-}
 function {
   id: 0x9bf1212e
   return_type_id: 0x6720d32f
@@ -332028,13 +331979,6 @@ function {
   return_type_id: 0x6720d32f
   parameter_id: 0x19c40476
 }
-function {
-  id: 0x9bf1ecfb
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x1582ab06
-  parameter_id: 0x2170d06d
-}
 function {
   id: 0x9bf23070
   return_type_id: 0x6720d32f
@@ -332211,13 +332155,6 @@ function {
   parameter_id: 0x1d19a9d5
   parameter_id: 0x92233392
 }
-function {
-  id: 0x9bff7f15
-  return_type_id: 0x6720d32f
-  parameter_id: 0x18bd6530
-  parameter_id: 0x16f99983
-  parameter_id: 0x2c8c423d
-}
 function {
   id: 0x9c00c8ec
   return_type_id: 0x6720d32f
@@ -349251,15 +349188,6 @@ elf_symbol {
   type_id: 0x9bec16df
   full_name: "__traceiter_android_vh_alloc_oem_binder_struct"
 }
-elf_symbol {
-  id: 0x4b794037
-  name: "__traceiter_android_vh_alloc_pages_end"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x3622b999
-  type_id: 0x9a878ec0
-  full_name: "__traceiter_android_vh_alloc_pages_end"
-}
 elf_symbol {
   id: 0x33f9278c
   name: "__traceiter_android_vh_alloc_pages_failure_bypass"
@@ -349287,33 +349215,6 @@ elf_symbol {
   type_id: 0x98788d90
   full_name: "__traceiter_android_vh_alloc_pages_slowpath"
 }
-elf_symbol {
-  id: 0xf3650061
-  name: "__traceiter_android_vh_alloc_pages_slowpath_end"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0xf3ad466f
-  type_id: 0x9b65d49e
-  full_name: "__traceiter_android_vh_alloc_pages_slowpath_end"
-}
-elf_symbol {
-  id: 0x23d552d4
-  name: "__traceiter_android_vh_alloc_pages_slowpath_start"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x2665cc80
-  type_id: 0x9b17d9b6
-  full_name: "__traceiter_android_vh_alloc_pages_slowpath_start"
-}
-elf_symbol {
-  id: 0x2bc6c2c3
-  name: "__traceiter_android_vh_alloc_pages_start"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0xccb286e6
-  type_id: 0x9b17d9b6
-  full_name: "__traceiter_android_vh_alloc_pages_start"
-}
 elf_symbol {
   id: 0x9c240ef1
   name: "__traceiter_android_vh_alloc_uid"
@@ -350106,15 +350007,6 @@ elf_symbol {
   type_id: 0x9a35263f
   full_name: "__traceiter_android_vh_do_futex"
 }
-elf_symbol {
-  id: 0x707cf8c6
-  name: "__traceiter_android_vh_do_new_mount_fc"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x35f39d7d
-  type_id: 0x9bff7f15
-  full_name: "__traceiter_android_vh_do_new_mount_fc"
-}
 elf_symbol {
   id: 0xea6452e1
   name: "__traceiter_android_vh_do_read_fault"
@@ -350322,15 +350214,6 @@ elf_symbol {
   type_id: 0x9bf4cdf9
   full_name: "__traceiter_android_vh_filemap_get_folio"
 }
-elf_symbol {
-  id: 0x2c39dada
-  name: "__traceiter_android_vh_filemap_get_folio_end"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x12b758f8
-  type_id: 0x9bf1ecfb
-  full_name: "__traceiter_android_vh_filemap_get_folio_end"
-}
 elf_symbol {
   id: 0xb7d91f76
   name: "__traceiter_android_vh_filemap_map_pages"
@@ -351042,24 +350925,6 @@ elf_symbol {
   type_id: 0x9b883042
   full_name: "__traceiter_android_vh_mm_compaction_end"
 }
-elf_symbol {
-  id: 0xecbe4183
-  name: "__traceiter_android_vh_mm_free"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x45d8a813
-  type_id: 0x9bc3617d
-  full_name: "__traceiter_android_vh_mm_free"
-}
-elf_symbol {
-  id: 0xc771047d
-  name: "__traceiter_android_vh_mm_init"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0xbc12a725
-  type_id: 0x9bc3617d
-  full_name: "__traceiter_android_vh_mm_init"
-}
 elf_symbol {
   id: 0x6f5c8275
   name: "__traceiter_android_vh_mm_kcompactd_cpu_online"
@@ -351411,15 +351276,6 @@ elf_symbol {
   type_id: 0x9b49a977
   full_name: "__traceiter_android_vh_ptype_head"
 }
-elf_symbol {
-  id: 0x358b7cff
-  name: "__traceiter_android_vh_put_super"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x55137777
-  type_id: 0x9b3403c3
-  full_name: "__traceiter_android_vh_put_super"
-}
 elf_symbol {
   id: 0x2c963d28
   name: "__traceiter_android_vh_queue_request_and_unlock"
@@ -351537,24 +351393,6 @@ elf_symbol {
   type_id: 0x9b5a0fe0
   full_name: "__traceiter_android_vh_regmap_update"
 }
-elf_symbol {
-  id: 0x4ac41e54
-  name: "__traceiter_android_vh_remove_mapping"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x2d2094ff
-  type_id: 0x9bf05906
-  full_name: "__traceiter_android_vh_remove_mapping"
-}
-elf_symbol {
-  id: 0xda7f66bc
-  name: "__traceiter_android_vh_remove_mapping_failed"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0xe58a89ce
-  type_id: 0x9bf05906
-  full_name: "__traceiter_android_vh_remove_mapping_failed"
-}
 elf_symbol {
   id: 0xf10ce16f
   name: "__traceiter_android_vh_resume_end"
@@ -352185,15 +352023,6 @@ elf_symbol {
   type_id: 0x9be67f35
   full_name: "__traceiter_android_vh_task_blocks_on_rtmutex"
 }
-elf_symbol {
-  id: 0x678eeaa9
-  name: "__traceiter_android_vh_task_mem"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x48cfcc61
-  type_id: 0x9bb0e048
-  full_name: "__traceiter_android_vh_task_mem"
-}
 elf_symbol {
   id: 0x592647e4
   name: "__traceiter_android_vh_tcp_rtt_estimator"
@@ -353040,15 +352869,6 @@ elf_symbol {
   type_id: 0x9bbf11b2
   full_name: "__traceiter_rpm_suspend"
 }
-elf_symbol {
-  id: 0x1e6a859e
-  name: "__traceiter_rss_stat"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x3d9d1e40
-  type_id: 0x9bd95971
-  full_name: "__traceiter_rss_stat"
-}
 elf_symbol {
   id: 0x2c11fea4
   name: "__traceiter_rwmmio_post_read"
@@ -354831,15 +354651,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_alloc_oem_binder_struct"
 }
-elf_symbol {
-  id: 0x7fdbf255
-  name: "__tracepoint_android_vh_alloc_pages_end"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xba6ec67f
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_alloc_pages_end"
-}
 elf_symbol {
   id: 0x37776872
   name: "__tracepoint_android_vh_alloc_pages_failure_bypass"
@@ -354867,33 +354678,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_alloc_pages_slowpath"
 }
-elf_symbol {
-  id: 0xff9480af
-  name: "__tracepoint_android_vh_alloc_pages_slowpath_end"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xe49b2157
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_alloc_pages_slowpath_end"
-}
-elf_symbol {
-  id: 0xa47162fe
-  name: "__tracepoint_android_vh_alloc_pages_slowpath_start"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x8ae9ae07
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_alloc_pages_slowpath_start"
-}
-elf_symbol {
-  id: 0x25e8e725
-  name: "__tracepoint_android_vh_alloc_pages_start"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xe9da2983
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_alloc_pages_start"
-}
 elf_symbol {
   id: 0xea695d5b
   name: "__tracepoint_android_vh_alloc_uid"
@@ -355686,15 +355470,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_do_futex"
 }
-elf_symbol {
-  id: 0xb75a1654
-  name: "__tracepoint_android_vh_do_new_mount_fc"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x5a168fc3
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_do_new_mount_fc"
-}
 elf_symbol {
   id: 0xff7bbbff
   name: "__tracepoint_android_vh_do_read_fault"
@@ -355902,15 +355677,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_filemap_get_folio"
 }
-elf_symbol {
-  id: 0x7b27aae8
-  name: "__tracepoint_android_vh_filemap_get_folio_end"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x6fb1cb00
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_filemap_get_folio_end"
-}
 elf_symbol {
   id: 0xb34d9200
   name: "__tracepoint_android_vh_filemap_map_pages"
@@ -356622,24 +356388,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_mm_compaction_end"
 }
-elf_symbol {
-  id: 0xf2ddbe8d
-  name: "__tracepoint_android_vh_mm_free"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xb63fd89e
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_mm_free"
-}
-elf_symbol {
-  id: 0xdae39903
-  name: "__tracepoint_android_vh_mm_init"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x0beb532e
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_mm_init"
-}
 elf_symbol {
   id: 0x0f593caf
   name: "__tracepoint_android_vh_mm_kcompactd_cpu_online"
@@ -356991,15 +356739,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_ptype_head"
 }
-elf_symbol {
-  id: 0x428834b1
-  name: "__tracepoint_android_vh_put_super"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x13d8191f
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_put_super"
-}
 elf_symbol {
   id: 0xf04b1c62
   name: "__tracepoint_android_vh_queue_request_and_unlock"
@@ -357117,24 +356856,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_regmap_update"
 }
-elf_symbol {
-  id: 0x8a5eba26
-  name: "__tracepoint_android_vh_remove_mapping"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xecf819e7
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_remove_mapping"
-}
-elf_symbol {
-  id: 0xa6c678f2
-  name: "__tracepoint_android_vh_remove_mapping_failed"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xdf0b8611
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_remove_mapping_failed"
-}
 elf_symbol {
   id: 0xd6904c6d
   name: "__tracepoint_android_vh_resume_end"
@@ -357765,15 +357486,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_android_vh_task_blocks_on_rtmutex"
 }
-elf_symbol {
-  id: 0xba9cb6ff
-  name: "__tracepoint_android_vh_task_mem"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0xa262475e
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_android_vh_task_mem"
-}
 elf_symbol {
   id: 0x3165f55a
   name: "__tracepoint_android_vh_tcp_rtt_estimator"
@@ -358620,15 +358332,6 @@ elf_symbol {
   type_id: 0x18ccbd2c
   full_name: "__tracepoint_rpm_suspend"
 }
-elf_symbol {
-  id: 0x3949a6b4
-  name: "__tracepoint_rss_stat"
-  is_defined: true
-  symbol_type: OBJECT
-  crc: 0x00e31afb
-  type_id: 0x18ccbd2c
-  full_name: "__tracepoint_rss_stat"
-}
 elf_symbol {
   id: 0x009c40ce
   name: "__tracepoint_rwmmio_post_read"
@@ -417237,24 +416940,6 @@ elf_symbol {
   type_id: 0x12d17df8
   full_name: "usb_wakeup_notification"
 }
-elf_symbol {
-  id: 0x03cd0310
-  name: "usbnet_cdc_bind"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x58bb6803
-  type_id: 0x9e768332
-  full_name: "usbnet_cdc_bind"
-}
-elf_symbol {
-  id: 0x27d4db38
-  name: "usbnet_cdc_status"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x64358e1e
-  type_id: 0x13eeeb77
-  full_name: "usbnet_cdc_status"
-}
 elf_symbol {
   id: 0x1227900d
   name: "usbnet_cdc_unbind"
@@ -417363,15 +417048,6 @@ elf_symbol {
   type_id: 0x12448959
   full_name: "usbnet_link_change"
 }
-elf_symbol {
-  id: 0x4a16e4d8
-  name: "usbnet_manage_power"
-  is_defined: true
-  symbol_type: FUNCTION
-  crc: 0x93c58c3a
-  type_id: 0x9f6e1a01
-  full_name: "usbnet_manage_power"
-}
 elf_symbol {
   id: 0x6c854cb3
   name: "usbnet_nway_reset"
@@ -422902,13 +422578,9 @@ interface {
   symbol_id: 0xf78be1c5
   symbol_id: 0x666f7184
   symbol_id: 0xd0b4a794
-  symbol_id: 0x4b794037
   symbol_id: 0x33f9278c
   symbol_id: 0x2e343246
   symbol_id: 0xa6e329cd
-  symbol_id: 0xf3650061
-  symbol_id: 0x23d552d4
-  symbol_id: 0x2bc6c2c3
   symbol_id: 0x9c240ef1
   symbol_id: 0x505f4289
   symbol_id: 0xd5431a65
@@ -422997,7 +422669,6 @@ interface {
   symbol_id: 0xb7d9e03a
   symbol_id: 0xada127d2
   symbol_id: 0xd593b3ef
-  symbol_id: 0x707cf8c6
   symbol_id: 0xea6452e1
   symbol_id: 0x1cc3aec5
   symbol_id: 0xb12728da
@@ -423021,7 +422692,6 @@ interface {
   symbol_id: 0x93a4717b
   symbol_id: 0x6091a763
   symbol_id: 0x49c69e22
-  symbol_id: 0x2c39dada
   symbol_id: 0xb7d91f76
   symbol_id: 0x4eda1196
   symbol_id: 0x158bf9d3
@@ -423101,8 +422771,6 @@ interface {
   symbol_id: 0x49cebcae
   symbol_id: 0x22de652b
   symbol_id: 0xf928bf8a
-  symbol_id: 0xecbe4183
-  symbol_id: 0xc771047d
   symbol_id: 0x6f5c8275
   symbol_id: 0xe44dacb1
   symbol_id: 0xb6126b05
@@ -423142,7 +422810,6 @@ interface {
   symbol_id: 0x574e7b58
   symbol_id: 0xf2c39651
   symbol_id: 0x93303c51
-  symbol_id: 0x358b7cff
   symbol_id: 0x2c963d28
   symbol_id: 0x3a545b61
   symbol_id: 0x96662dde
@@ -423156,8 +422823,6 @@ interface {
   symbol_id: 0xbb291efa
   symbol_id: 0xf2957eae
   symbol_id: 0xe2d75052
-  symbol_id: 0x4ac41e54
-  symbol_id: 0xda7f66bc
   symbol_id: 0xf10ce16f
   symbol_id: 0xdd585356
   symbol_id: 0xa8cef421
@@ -423228,7 +422893,6 @@ interface {
   symbol_id: 0x2ecf85e9
   symbol_id: 0x34a01a22
   symbol_id: 0xdd9dd67b
-  symbol_id: 0x678eeaa9
   symbol_id: 0x592647e4
   symbol_id: 0x48f0cf25
   symbol_id: 0x40d428a7
@@ -423323,7 +422987,6 @@ interface {
   symbol_id: 0x189202e9
   symbol_id: 0x04c5b642
   symbol_id: 0x88744d70
-  symbol_id: 0x1e6a859e
   symbol_id: 0x2c11fea4
   symbol_id: 0x45e62871
   symbol_id: 0x85a11e58
@@ -423522,13 +423185,9 @@ interface {
   symbol_id: 0x7ea1f05b
   symbol_id: 0x74cce3b2
   symbol_id: 0xd7f9868e
-  symbol_id: 0x7fdbf255
   symbol_id: 0x37776872
   symbol_id: 0xfc735654
   symbol_id: 0x1ebb872f
-  symbol_id: 0xff9480af
-  symbol_id: 0xa47162fe
-  symbol_id: 0x25e8e725
   symbol_id: 0xea695d5b
   symbol_id: 0xd82141b7
   symbol_id: 0xe520cb67
@@ -423617,7 +423276,6 @@ interface {
   symbol_id: 0x066e59a8
   symbol_id: 0x4db9ff7c
   symbol_id: 0x9fe99d05
-  symbol_id: 0xb75a1654
   symbol_id: 0xff7bbbff
   symbol_id: 0x82ce823f
   symbol_id: 0x474d211c
@@ -423641,7 +423299,6 @@ interface {
   symbol_id: 0x50a83025
   symbol_id: 0x0c03d499
   symbol_id: 0x6d970e8c
-  symbol_id: 0x7b27aae8
   symbol_id: 0xb34d9200
   symbol_id: 0x223c9b64
   symbol_id: 0xf61927fd
@@ -423721,8 +423378,6 @@ interface {
   symbol_id: 0xf2890610
   symbol_id: 0xa3d5f70d
   symbol_id: 0x72c79d80
-  symbol_id: 0xf2ddbe8d
-  symbol_id: 0xdae39903
   symbol_id: 0x0f593caf
   symbol_id: 0xb6da564f
   symbol_id: 0xa1c3d037
@@ -423762,7 +423417,6 @@ interface {
   symbol_id: 0x61f8c8ae
   symbol_id: 0x0e92ee53
   symbol_id: 0xb0c197a3
-  symbol_id: 0x428834b1
   symbol_id: 0xf04b1c62
   symbol_id: 0x811d5fab
   symbol_id: 0x2d9a331c
@@ -423776,8 +423430,6 @@ interface {
   symbol_id: 0x1fe1da5c
   symbol_id: 0x57c5897c
   symbol_id: 0x13b2fb38
-  symbol_id: 0x8a5eba26
-  symbol_id: 0xa6c678f2
   symbol_id: 0xd6904c6d
   symbol_id: 0x3cdecaec
   symbol_id: 0xcad24ea3
@@ -423848,7 +423500,6 @@ interface {
   symbol_id: 0xefb9e5a3
   symbol_id: 0x3fe0157c
   symbol_id: 0xe5bf742d
-  symbol_id: 0xba9cb6ff
   symbol_id: 0x3165f55a
   symbol_id: 0x4ef2c337
   symbol_id: 0xafd4e4b1
@@ -423943,7 +423594,6 @@ interface {
   symbol_id: 0xfbd9814b
   symbol_id: 0x7c591834
   symbol_id: 0x636e234e
-  symbol_id: 0x3949a6b4
   symbol_id: 0x009c40ce
   symbol_id: 0x02473ce7
   symbol_id: 0x65f99dde
@@ -430457,8 +430107,6 @@ interface {
   symbol_id: 0xcac8f190
   symbol_id: 0x9ea1c58f
   symbol_id: 0x140d9164
-  symbol_id: 0x03cd0310
-  symbol_id: 0x27d4db38
   symbol_id: 0x1227900d
   symbol_id: 0x8189acd0
   symbol_id: 0xbbf007f4
@@ -430471,7 +430119,6 @@ interface {
   symbol_id: 0x3f51a906
   symbol_id: 0x6515c416
   symbol_id: 0x19d80976
-  symbol_id: 0x4a16e4d8
   symbol_id: 0x6c854cb3
   symbol_id: 0xcd021e80
   symbol_id: 0x147e4ad2
diff --git a/aosp/android/abi_gki_aarch64_galaxy b/aosp/android/abi_gki_aarch64_galaxy
index 02bf4a491e6d..064f2b307565 100644
--- a/aosp/android/abi_gki_aarch64_galaxy
+++ b/aosp/aosp/android/abi_gki_aarch64_galaxy
@@ -342,8 +342,6 @@
   __traceiter_android_vh_madvise_pageout_swap_entry
   __traceiter_android_vh_madvise_swapin_walk_pmd_entry
   __traceiter_android_vh_meminfo_cache_adjust
-  __traceiter_android_vh_mm_free
-  __traceiter_android_vh_mm_init
   __traceiter_android_vh_mutex_wait_finish
   __traceiter_android_vh_mutex_wait_start
   __traceiter_android_vh_process_madvise_end
@@ -359,7 +357,6 @@
   __traceiter_android_vh_si_mem_available_adjust
   __traceiter_android_vh_si_meminfo_adjust
   __traceiter_android_vh_smaps_pte_entry
-  __traceiter_android_vh_task_mem
   __traceiter_android_vh_try_to_freeze_todo
   __traceiter_android_vh_try_to_freeze_todo_unfrozen
   __traceiter_android_vh_watchdog_timer_softlockup
@@ -372,7 +369,6 @@
   __traceiter_irq_handler_entry
   __traceiter_irq_handler_exit
   __traceiter_kfree_skb
-  __traceiter_rss_stat
   __traceiter_sched_switch
   __traceiter_workqueue_execute_start
   __tracepoint_android_rvh_arm64_serror_panic
@@ -389,8 +385,6 @@
   __tracepoint_android_vh_madvise_pageout_swap_entry
   __tracepoint_android_vh_madvise_swapin_walk_pmd_entry
   __tracepoint_android_vh_meminfo_cache_adjust
-  __tracepoint_android_vh_mm_free
-  __tracepoint_android_vh_mm_init
   __tracepoint_android_vh_mutex_wait_finish
   __tracepoint_android_vh_mutex_wait_start
   __tracepoint_android_vh_process_madvise_end
@@ -406,7 +400,6 @@
   __tracepoint_android_vh_si_mem_available_adjust
   __tracepoint_android_vh_si_meminfo_adjust
   __tracepoint_android_vh_smaps_pte_entry
-  __tracepoint_android_vh_task_mem
   __tracepoint_android_vh_try_to_freeze_todo
   __tracepoint_android_vh_try_to_freeze_todo_unfrozen
   __tracepoint_android_vh_watchdog_timer_softlockup
@@ -421,7 +414,6 @@
   __tracepoint_kfree_skb
   tracepoint_probe_register
   tracepoint_probe_unregister
-  __tracepoint_rss_stat
   __tracepoint_sched_switch
   __tracepoint_workqueue_execute_start
   unregister_console
diff --git a/aosp/android/abi_gki_aarch64_wonder b/aosp/android/abi_gki_aarch64_wonder
deleted file mode 100644
index 967ebcacca20..000000000000
--- a/aosp/android/abi_gki_aarch64_wonder
+++ /dev/null
@@ -1,338 +0,0 @@
-[abi_symbol_list]
-# commonly used symbols
-  __alloc_skb
-  alloc_workqueue
-  alt_cb_patch_nops
-  bcmp
-  bpf_trace_run1
-  bpf_trace_run2
-  bpf_trace_run3
-  bpf_trace_run4
-  bpf_trace_run5
-  bpf_trace_run6
-  cancel_delayed_work
-  cancel_delayed_work_sync
-  cancel_work_sync
-  consume_skb
-  cpu_number
-  __cpu_online_mask
-  crc32_le
-  csum_partial
-  debugfs_create_dir
-  delayed_work_timer_fn
-  del_timer
-  del_timer_sync
-  destroy_workqueue
-  dev_addr_mod
-  dev_close
-  _dev_err
-  __dev_kfree_skb_any
-  __dev_queue_xmit
-  eth_type_trans
-  flush_delayed_work
-  fortify_panic
-  get_random_bytes
-  init_net
-  init_timer_key
-  jiffies
-  jiffies_to_msecs
-  kfree
-  kfree_sensitive
-  kfree_skb_reason
-  __kmalloc
-  kmalloc_caches
-  kmalloc_trace
-  kmemdup
-  kvfree_call_rcu
-  __list_add_valid
-  __list_del_entry_valid
-  __local_bh_enable_ip
-  memcpy
-  memset
-  mod_timer
-  module_layout
-  __msecs_to_jiffies
-  __mutex_init
-  mutex_lock
-  mutex_unlock
-  __netdev_alloc_skb
-  netif_receive_skb
-  netif_rx
-  nla_memcpy
-  __nla_parse
-  nla_put
-  param_ops_bool
-  param_ops_charp
-  param_ops_int
-  perf_trace_buf_alloc
-  perf_trace_run_bpf_submit
-  preempt_schedule_notrace
-  _printk
-  __pskb_pull_tail
-  __put_task_struct
-  queue_delayed_work_on
-  queue_work_on
-  ___ratelimit
-  _raw_spin_lock
-  _raw_spin_lock_bh
-  _raw_spin_lock_irqsave
-  _raw_spin_unlock
-  _raw_spin_unlock_bh
-  _raw_spin_unlock_irqrestore
-  __rcu_read_lock
-  __rcu_read_unlock
-  register_netdevice_notifier
-  rtnl_is_locked
-  rtnl_lock
-  rtnl_unlock
-  skb_copy_bits
-  skb_pull
-  skb_push
-  skb_put
-  skb_trim
-  snprintf
-  __stack_chk_fail
-  strcmp
-  strcpy
-  strlen
-  strnlen
-  strscpy
-  __sw_hweight16
-  __sw_hweight32
-  __sw_hweight8
-  synchronize_net
-  synchronize_rcu
-  system_wq
-  trace_event_buffer_commit
-  trace_event_buffer_reserve
-  trace_event_printf
-  trace_event_raw_init
-  trace_event_reg
-  trace_handle_return
-  trace_raw_output_prep
-  __trace_trigger_soft_disabled
-  unregister_netdevice_notifier
-  unregister_netdevice_queue
-  __warn_printk
-
-# required by cfg80211.ko
-  bpf_trace_run10
-  bpf_trace_run7
-  bpf_trace_run8
-  __class_register
-  class_unregister
-  _ctype
-  debugfs_remove
-  debugfs_rename
-  __dev_change_net_namespace
-  __dev_get_by_index
-  dev_get_by_index
-  device_add
-  device_del
-  device_initialize
-  device_rename
-  dev_set_name
-  do_trace_netlink_extack
-  finish_wait
-  flush_work
-  gcd
-  genlmsg_multicast_allns
-  genlmsg_put
-  genl_register_family
-  genl_unregister_family
-  get_net_ns_by_fd
-  get_net_ns_by_pid
-  inet_csk_get_port
-  init_uts_ns
-  init_wait_entry
-  __init_waitqueue_head
-  kasan_flag_enabled
-  key_create_or_update
-  key_put
-  keyring_alloc
-  kobject_uevent_env
-  ktime_get_coarse_with_offset
-  memcmp
-  mod_delayed_work_on
-  netdev_err
-  netlink_broadcast
-  netlink_register_notifier
-  netlink_unicast
-  netlink_unregister_notifier
-  net_ns_type_operations
-  nla_find
-  nla_put_64bit
-  nla_reserve
-  __nla_validate
-  of_find_property
-  of_prop_next_u32
-  platform_device_register_full
-  platform_device_unregister
-  prepare_to_wait_event
-  put_device
-  __put_net
-  _raw_spin_lock_irq
-  _raw_spin_unlock_irq
-  rb_erase
-  rb_insert_color
-  refcount_warn_saturate
-  register_netdevice
-  register_pernet_device
-  release_firmware
-  request_firmware
-  request_firmware_nowait
-  rfkill_alloc
-  rfkill_blocked
-  rfkill_destroy
-  rfkill_register
-  rfkill_resume_polling
-  rfkill_set_hw_state_reason
-  rfkill_unregister
-  schedule
-  scnprintf
-  skb_add_rx_frag
-  __sock_create
-  sock_release
-  sprintf
-  sscanf
-  sysfs_create_link
-  sysfs_remove_link
-  system_power_efficient_wq
-  system_unbound_wq
-  trace_print_array_seq
-  unregister_pernet_device
-  verify_pkcs7_signature
-  __wake_up
-  wireless_nlevent_flush
-
-# required by mac80211.ko
-  alloc_netdev_mqs
-  __alloc_percpu_gfp
-  arc4_crypt
-  arc4_setkey
-  bitmap_free
-  bitmap_zalloc
-  call_rcu
-  __cpu_possible_mask
-  crc32_be
-  crypto_aead_decrypt
-  crypto_aead_encrypt
-  crypto_aead_setauthsize
-  crypto_aead_setkey
-  crypto_alloc_aead
-  crypto_alloc_shash
-  crypto_alloc_skcipher
-  crypto_destroy_tfm
-  __crypto_memneq
-  crypto_shash_digest
-  crypto_shash_finup
-  crypto_shash_setkey
-  crypto_shash_update
-  crypto_skcipher_decrypt
-  crypto_skcipher_encrypt
-  crypto_skcipher_setkey
-  __crypto_xor
-  dev_alloc_name
-  dev_fetch_sw_netstats
-  _dev_info
-  _dev_printk
-  _dev_warn
-  ether_setup
-  eth_mac_addr
-  ethtool_op_get_link
-  _find_first_bit
-  _find_next_bit
-  __flush_workqueue
-  free_netdev
-  free_percpu
-  get_random_u16
-  __hw_addr_init
-  __hw_addr_sync
-  __hw_addr_unsync
-  idr_alloc
-  idr_destroy
-  idr_find
-  idr_for_each
-  idr_get_next
-  idr_remove
-  kernel_param_lock
-  kernel_param_unlock
-  kfree_skb_list_reason
-  kmalloc_large
-  ktime_get
-  ktime_get_seconds
-  ktime_get_with_offset
-  kvfree
-  kvmalloc_node
-  memmove
-  napi_gro_receive
-  netdev_info
-  netdev_set_default_ethtool_ops
-  netif_carrier_off
-  netif_carrier_on
-  netif_receive_skb_list
-  netif_tx_stop_all_queues
-  netif_tx_wake_queue
-  net_ratelimit
-  nr_cpu_ids
-  __per_cpu_offset
-  preempt_schedule
-  pskb_expand_head
-  ___pskb_trim
-  rcu_barrier
-  register_inet6addr_notifier
-  register_inetaddr_notifier
-  rhashtable_free_and_destroy
-  rhashtable_insert_slow
-  rhltable_init
-  __rht_bucket_nested
-  rht_bucket_nested
-  rht_bucket_nested_insert
-  round_jiffies
-  round_jiffies_relative
-  round_jiffies_up
-  sg_init_one
-  sg_init_table
-  skb_checksum_help
-  skb_clone
-  skb_clone_sk
-  skb_complete_wifi_ack
-  skb_copy
-  skb_copy_expand
-  skb_dequeue
-  skb_ensure_writable
-  __skb_get_hash
-  __skb_gso_segment
-  skb_queue_head
-  skb_queue_purge
-  skb_queue_tail
-  strncpy
-  __sw_hweight64
-  system_freezable_wq
-  tasklet_kill
-  __tasklet_schedule
-  tasklet_setup
-  tasklet_unlock_wait
-  unregister_inet6addr_notifier
-  unregister_inetaddr_notifier
-  unregister_netdevice_many
-  __usecs_to_jiffies
-
-# required by wonder.ko
-  __auxiliary_driver_register
-  auxiliary_driver_unregister
-  debugfs_create_file
-  debugfs_lookup_and_remove
-  dev_get_by_name
-  kstrtobool_from_user
-  netdev_rx_handler_register
-  netdev_rx_handler_unregister
-  noop_llseek
-  seq_lseek
-  seq_printf
-  seq_read
-  simple_open
-  single_open
-  single_release
-  vfree
-  vzalloc
diff --git a/aosp/android/abi_gki_aarch64_xiaomi b/aosp/android/abi_gki_aarch64_xiaomi
index 94d423257eaa..02cbfd9301f4 100644
--- a/aosp/android/abi_gki_aarch64_xiaomi
+++ b/aosp/aosp/android/abi_gki_aarch64_xiaomi
@@ -516,17 +516,6 @@ __traceiter_android_vh_do_folio_trylock
 __tracepoint_android_vh_do_folio_trylock
 reclaim_pages
 
-__traceiter_android_vh_remove_mapping
-__traceiter_android_vh_remove_mapping_failed
-__traceiter_android_vh_filemap_get_folio_end
-__traceiter_android_vh_do_new_mount_fc
-__traceiter_android_vh_put_super
-__tracepoint_android_vh_remove_mapping
-__tracepoint_android_vh_remove_mapping_failed
-__tracepoint_android_vh_filemap_get_folio_end
-__tracepoint_android_vh_do_new_mount_fc
-__tracepoint_android_vh_put_super
-
 #required by mi_iod_debug.ko
 page_swap_info
 
@@ -557,13 +546,3 @@ __tracepoint_android_vh_pcp_alloc_factor_adjust
 # required by mi cma pool
 cma_mutex
 
-# required by low_lat_mem dfx
-__traceiter_android_vh_alloc_pages_start
-__tracepoint_android_vh_alloc_pages_start
-__traceiter_android_vh_alloc_pages_end
-__tracepoint_android_vh_alloc_pages_end
-__traceiter_android_vh_alloc_pages_slowpath_start
-__tracepoint_android_vh_alloc_pages_slowpath_start
-__traceiter_android_vh_alloc_pages_slowpath_end
-__tracepoint_android_vh_alloc_pages_slowpath_end
-
diff --git a/aosp/android/abi_gki_aarch64_zebra b/aosp/android/abi_gki_aarch64_zebra
index 2971176a8493..785354d942cb 100644
--- a/aosp/android/abi_gki_aarch64_zebra
+++ b/aosp/aosp/android/abi_gki_aarch64_zebra
@@ -32,7 +32,3 @@
   v4l2_pipeline_pm_get
   vb2_video_unregister_device
   v4l2_get_link_freq
-#required by ethernet driver
-  usbnet_manage_power
-  usbnet_cdc_status
-  usbnet_cdc_bind
diff --git a/aosp/android/abi_gki_protected_exports_aarch64 b/aosp/android/abi_gki_protected_exports_aarch64
index 620de6499ecc..5668a34809e1 100644
--- a/aosp/android/abi_gki_protected_exports_aarch64
+++ b/aosp/aosp/android/abi_gki_protected_exports_aarch64
@@ -209,7 +209,6 @@ nfc_get_local_general_bytes
 nfc_proto_register
 nfc_proto_unregister
 nfc_register_device
-nfc_remove_device
 nfc_remove_se
 nfc_se_connectivity
 nfc_se_transaction
@@ -221,7 +220,6 @@ nfc_tm_activated
 nfc_tm_data_received
 nfc_tm_deactivated
 nfc_unregister_device
-nfc_unregister_rfkill
 nfc_vendor_cmd_reply
 of_can_transceiver
 open_candev
diff --git a/aosp/arch/arm64/configs/gki_defconfig b/aosp/arch/arm64/configs/gki_defconfig
index 06e28c0f31ec..4e437f9c956e 100644
--- a/aosp/arch/arm64/configs/gki_defconfig
+++ b/aosp/aosp/arch/arm64/configs/gki_defconfig
@@ -123,6 +123,7 @@ CONFIG_CMA=y
 CONFIG_CMA_DEBUGFS=y
 CONFIG_CMA_AREAS=32
 # CONFIG_ZONE_DMA is not set
+CONFIG_MEMFD_ASHMEM_SHIM=y
 CONFIG_ANON_VMA_NAME=y
 CONFIG_USERFAULTFD=y
 CONFIG_LRU_GEN=y
diff --git a/aosp/arch/arm64/configs/wonder.fragment b/aosp/arch/arm64/configs/wonder.fragment
deleted file mode 100644
index 78d0283f73c3..000000000000
--- a/aosp/arch/arm64/configs/wonder.fragment
+++ /dev/null
@@ -1,6 +0,0 @@
-# CONFIG_MODULE_SIG_ALL is not set
-CONFIG_CFG80211=m
-CONFIG_NL80211_TESTMODE=y
-CONFIG_CFG80211_CERTIFICATION_ONUS=y
-CONFIG_CFG80211_REG_CELLULAR_HINTS=y
-CONFIG_MAC80211=m
diff --git a/aosp/arch/arm64/kernel/patch-scs.c b/aosp/arch/arm64/kernel/patch-scs.c
index a1fe4b4ff591..1b3da02d5b74 100644
--- a/aosp/arch/arm64/kernel/patch-scs.c
+++ b/aosp/aosp/arch/arm64/kernel/patch-scs.c
@@ -130,8 +130,7 @@ struct eh_frame {
 
 static int noinstr scs_handle_fde_frame(const struct eh_frame *frame,
 					bool fde_has_augmentation_data,
-					int code_alignment_factor,
-					bool dry_run)
+					int code_alignment_factor)
 {
 	int size = frame->size - offsetof(struct eh_frame, opcodes) + 4;
 	u64 loc = (u64)offset_to_ptr(&frame->initial_loc);
@@ -185,8 +184,7 @@ static int noinstr scs_handle_fde_frame(const struct eh_frame *frame,
 			break;
 
 		case DW_CFA_negate_ra_state:
-			if (!dry_run)
-				scs_patch_loc(loc - 4);
+			scs_patch_loc(loc - 4);
 			break;
 
 		case 0x40 ... 0x7f:
@@ -237,12 +235,9 @@ int noinstr scs_patch(const u8 eh_frame[], int size)
 		} else {
 			ret = scs_handle_fde_frame(frame,
 						   fde_has_augmentation_data,
-						   code_alignment_factor,
-						   true);
+						   code_alignment_factor);
 			if (ret)
 				return ret;
-			scs_handle_fde_frame(frame, fde_has_augmentation_data,
-					     code_alignment_factor, false);
 		}
 
 		p += sizeof(frame->size) + frame->size;
diff --git a/aosp/arch/x86/configs/gki_defconfig b/aosp/arch/x86/configs/gki_defconfig
index 5df374201224..2474e269377a 100644
--- a/aosp/arch/x86/configs/gki_defconfig
+++ b/aosp/aosp/arch/x86/configs/gki_defconfig
@@ -119,6 +119,7 @@ CONFIG_CMA=y
 CONFIG_CMA_DEBUGFS=y
 CONFIG_CMA_AREAS=16
 # CONFIG_ZONE_DMA is not set
+CONFIG_MEMFD_ASHMEM_SHIM=y
 CONFIG_ANON_VMA_NAME=y
 CONFIG_USERFAULTFD=y
 CONFIG_LRU_GEN=y
@@ -236,6 +237,7 @@ CONFIG_IP6_NF_FILTER=y
 CONFIG_IP6_NF_TARGET_REJECT=y
 CONFIG_IP6_NF_MANGLE=y
 CONFIG_IP6_NF_RAW=y
+CONFIG_TIPC=m
 CONFIG_L2TP=m
 CONFIG_BRIDGE=y
 CONFIG_VLAN_8021Q=m
diff --git a/aosp/build.config.wonder b/aosp/build.config.wonder
deleted file mode 100644
index 725f7b360023..000000000000
--- a/aosp/build.config.wonder
+++ /dev/null
@@ -1,7 +0,0 @@
-. ${ROOT_DIR}/${KERNEL_DIR}/build.config.common
-. ${ROOT_DIR}/${KERNEL_DIR}/build.config.aarch64
-
-DEFCONFIG=wonder_gki_defconfig
-FRAGMENT_CONFIG=${KERNEL_DIR}/arch/arm64/configs/wonder.fragment
-PRE_DEFCONFIG_CMDS="KCONFIG_CONFIG=${ROOT_DIR}/${KERNEL_DIR}/arch/arm64/configs/${DEFCONFIG} ${ROOT_DIR}/${KERNEL_DIR}/scripts/kconfig/merge_config.sh -m -r ${ROOT_DIR}/${KERNEL_DIR}/arch/arm64/configs/gki_defconfig ${ROOT_DIR}/${FRAGMENT_CONFIG}"
-POST_DEFCONFIG_CMDS="rm ${ROOT_DIR}/${KERNEL_DIR}/arch/arm64/configs/${DEFCONFIG}"
diff --git a/aosp/drivers/android/vendor_hooks.c b/aosp/drivers/android/vendor_hooks.c
index 115329cb3771..5bb0a2a12994 100644
--- a/aosp/drivers/android/vendor_hooks.c
+++ b/aosp/aosp/drivers/android/vendor_hooks.c
@@ -66,7 +66,6 @@
 #include <trace/hooks/signal.h>
 #include <trace/hooks/psci.h>
 #include <trace/hooks/psi.h>
-#include <trace/hooks/blk.h>
 #include <trace/hooks/bl_hib.h>
 #include <trace/hooks/regmap.h>
 #include <trace/hooks/compaction.h>
@@ -81,7 +80,6 @@
  * Export tracepoints that act as a bare tracehook (ie: have no trace event
  * associated with them) to allow external modules to probe them.
  */
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_put_super);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_sk_alloc);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_sk_free);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_refrigerator);
@@ -489,12 +487,9 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_finish);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_evaluate_task_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_swapmem_gather_init);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_swapmem_gather_finish);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_new_mount_fc);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_read_fault);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_read);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_map_pages);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_remove_mapping);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_remove_mapping_failed);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mmc_blk_reset);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mmc_attach_sd);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_sdhci_get_cd);
@@ -504,7 +499,6 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_map_pages_range);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_vprintk_store);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_folio_referenced_check_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_calculate_totalreserve_pages);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_get_folio_end);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_drain_all_pages_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_hibernate_resume_state);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_pageset_update);
@@ -514,10 +508,3 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_create_worker);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_fas_gpu_qos_update_tracer);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_nr_pcp_alloc);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_pcp_alloc_factor_adjust);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_task_mem);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_free);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_init);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_alloc_pages_start);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_alloc_pages_end);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_alloc_pages_slowpath_start);
-EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_alloc_pages_slowpath_end);
diff --git a/aosp/fs/namespace.c b/aosp/fs/namespace.c
index 4d910be7cc9a..f22f76d9c22f 100644
--- a/aosp/fs/namespace.c
+++ b/aosp/aosp/fs/namespace.c
@@ -35,7 +35,6 @@
 
 #include "pnode.h"
 #include "internal.h"
-#include <trace/hooks/blk.h>
 
 /* Maximum number of mounts in a mount namespace */
 static unsigned int sysctl_mount_max __read_mostly = 100000;
@@ -3022,8 +3021,6 @@ static int do_new_mount_fc(struct fs_context *fc, struct path *mountpoint,
 	unlock_mount(mp);
 	if (error < 0)
 		mntput(mnt);
-	else
-		trace_android_vh_do_new_mount_fc(mountpoint, mnt);
 	return error;
 }
 
diff --git a/aosp/fs/proc/task_mmu.c b/aosp/fs/proc/task_mmu.c
index 37f6cefe07be..d6f37ed840e5 100644
--- a/aosp/fs/proc/task_mmu.c
+++ b/aosp/aosp/fs/proc/task_mmu.c
@@ -78,7 +78,6 @@ void task_mem(struct seq_file *m, struct mm_struct *mm)
 	SEQ_PUT_DEC(" kB\nVmSwap:\t", swap);
 	seq_puts(m, " kB\n");
 	hugetlb_report_usage(m, mm);
-	trace_android_vh_task_mem(m, mm);
 }
 #undef SEQ_PUT_DEC
 
diff --git a/aosp/fs/super.c b/aosp/fs/super.c
index a50cc4fdc523..904839eb7a3d 100644
--- a/aosp/fs/super.c
+++ b/aosp/aosp/fs/super.c
@@ -40,8 +40,6 @@
 #include <uapi/linux/mount.h>
 #include "internal.h"
 
-#include <trace/hooks/fs.h>
-
 static int thaw_super_locked(struct super_block *sb);
 
 static LIST_HEAD(super_blocks);
@@ -501,7 +499,6 @@ void generic_shutdown_super(struct super_block *sb)
 			sb->s_dio_done_wq = NULL;
 		}
 
-		trace_android_vh_put_super(sb);
 		if (sop->put_super)
 			sop->put_super(sb);
 
diff --git a/aosp/include/trace/hooks/blk.h b/aosp/include/trace/hooks/blk.h
deleted file mode 100644
index 0eda68c5a3be..000000000000
--- a/aosp/include/trace/hooks/blk.h
+++ /dev/null
@@ -1,22 +0,0 @@
-/* SPDX-License-Identifier: GPL-2.0 */
-#undef TRACE_SYSTEM
-#define TRACE_SYSTEM blk
-
-#undef TRACE_INCLUDE_PATH
-#define TRACE_INCLUDE_PATH trace/hooks
-
-#if !defined(_TRACE_HOOK_BLK_H) || defined(TRACE_HEADER_MULTI_READ)
-#define _TRACE_HOOK_BLK_H
-
-#include <trace/hooks/vendor_hooks.h>
-
-struct path;
-struct vfsmount;
-
-DECLARE_HOOK(android_vh_do_new_mount_fc,
-	TP_PROTO(struct path *mountpoint, struct vfsmount *mnt),
-	TP_ARGS(mountpoint, mnt));
-
-#endif /* _TRACE_HOOK_BLK_H */
-/* This part must be outside protection */
-#include <trace/define_trace.h>
diff --git a/aosp/include/trace/hooks/fs.h b/aosp/include/trace/hooks/fs.h
index 2ff10dcd84fa..bb8f177db5c1 100644
--- a/aosp/include/trace/hooks/fs.h
+++ b/aosp/aosp/include/trace/hooks/fs.h
@@ -10,10 +10,6 @@
 
 #include <trace/hooks/vendor_hooks.h>
 
-DECLARE_HOOK(android_vh_put_super,
-	TP_PROTO(struct super_block *sb),
-	TP_ARGS(sb));
-
 DECLARE_HOOK(android_vh_ep_create_wakeup_source,
 	TP_PROTO(char *name, int len),
 	TP_ARGS(name, len));
@@ -24,4 +20,4 @@ DECLARE_HOOK(android_vh_timerfd_create,
 #endif /* _TRACE_HOOK_FS_H */
 
 /* This part must be outside protection */
-#include <trace/define_trace.h>
+#include <trace/define_trace.h>
\ No newline at end of file
diff --git a/aosp/include/trace/hooks/mm.h b/aosp/include/trace/hooks/mm.h
index af67ded7de78..c02aa2ca1fe8 100644
--- a/aosp/include/trace/hooks/mm.h
+++ b/aosp/aosp/include/trace/hooks/mm.h
@@ -341,9 +341,6 @@ DECLARE_HOOK(android_vh_pageset_update,
 DECLARE_HOOK(android_vh_mempool_alloc_skip_wait,
 	TP_PROTO(gfp_t *gfp_flags, bool *skip_wait),
 	TP_ARGS(gfp_flags, skip_wait));
-DECLARE_HOOK(android_vh_filemap_get_folio_end,
-	TP_PROTO(struct address_space *mapping, struct folio *folio),
-	TP_ARGS(mapping, folio));
 DECLARE_HOOK(android_vh_nr_pcp_alloc,
 	TP_PROTO(struct per_cpu_pages *pcp, struct zone *zone,
 		unsigned long __percpu **pad, unsigned int order, int *batch),
@@ -353,30 +350,6 @@ DECLARE_HOOK(android_vh_pcp_alloc_factor_adjust,
 		struct per_cpu_pages *pcp, struct page *page, int migratetype,
 		unsigned int order),
 	TP_ARGS(zone, pad, pcp, page, migratetype, order));
-DECLARE_HOOK(android_vh_task_mem,
-	TP_PROTO(struct seq_file *m, struct mm_struct *mm),
-	TP_ARGS(m, mm));
-DECLARE_HOOK(android_vh_mm_free,
-	TP_PROTO(struct mm_struct *mm),
-	TP_ARGS(mm));
-DECLARE_HOOK(android_vh_mm_init,
-	TP_PROTO(struct mm_struct *mm),
-	TP_ARGS(mm));
-DECLARE_HOOK(android_vh_alloc_pages_start,
-	TP_PROTO(u64 *stime),
-	TP_ARGS(stime));
-DECLARE_HOOK(android_vh_alloc_pages_end,
-	TP_PROTO(unsigned int order, gfp_t alloc_gfp, u64 stime),
-	TP_ARGS(order, alloc_gfp, stime));
-DECLARE_HOOK(android_vh_alloc_pages_slowpath_start,
-	TP_PROTO(u64 *stime),
-	TP_ARGS(stime));
-DECLARE_HOOK(android_vh_alloc_pages_slowpath_end,
-	TP_PROTO(gfp_t *gfp_mask, unsigned int order, unsigned long alloc_start,
-		u64 stime, unsigned long did_some_progress,
-		unsigned long pages_reclaimed, int retry_loop_count),
-	TP_ARGS(gfp_mask, order, alloc_start, stime, did_some_progress,
-		pages_reclaimed, retry_loop_count));
 #endif /* _TRACE_HOOK_MM_H */
 
 /* This part must be outside protection */
diff --git a/aosp/include/trace/hooks/vmscan.h b/aosp/include/trace/hooks/vmscan.h
index f6f93a292f44..52eff72945b5 100644
--- a/aosp/include/trace/hooks/vmscan.h
+++ b/aosp/aosp/include/trace/hooks/vmscan.h
@@ -43,12 +43,6 @@ DECLARE_HOOK(android_vh_should_continue_reclaim,
 DECLARE_HOOK(android_vh_file_is_tiny_bypass,
 	TP_PROTO(bool file_is_tiny, bool *bypass),
 	TP_ARGS(file_is_tiny, bypass));
-DECLARE_HOOK(android_vh_remove_mapping,
-	TP_PROTO(struct address_space *mapping, struct folio *folio, bool reclaimed),
-	TP_ARGS(mapping, folio, reclaimed));
-DECLARE_HOOK(android_vh_remove_mapping_failed,
-	TP_PROTO(struct address_space *mapping, struct folio *folio, bool reclaimed),
-	TP_ARGS(mapping, folio, reclaimed));
 DECLARE_HOOK(android_vh_handle_folio_writeback,
 	TP_PROTO(struct folio *folio, bool *bypass),
 	TP_ARGS(folio, bypass));
diff --git a/aosp/kernel/fork.c b/aosp/kernel/fork.c
index 77643a8aaa43..3cdbf43930bc 100644
--- a/aosp/kernel/fork.c
+++ b/aosp/aosp/kernel/fork.c
@@ -114,7 +114,6 @@
 
 #undef CREATE_TRACE_POINTS
 #include <trace/hooks/sched.h>
-#include <trace/hooks/mm.h>
 /*
  * Minimum number of threads to boot the kernel
  */
@@ -889,7 +888,6 @@ void __mmdrop(struct mm_struct *mm)
 	put_user_ns(mm->user_ns);
 	mm_pasid_drop(mm);
 	kfree(mm->abi_extend);
-	trace_android_vh_mm_free(mm);
 	free_mm(mm);
 }
 EXPORT_SYMBOL_GPL(__mmdrop);
@@ -1268,7 +1266,6 @@ static struct mm_struct *mm_init(struct mm_struct *mm, struct task_struct *p,
 
 	mm->user_ns = get_user_ns(user_ns);
 	lru_gen_init_mm(mm);
-	trace_android_vh_mm_init(mm);
 	return mm;
 
 fail_nocontext:
diff --git a/aosp/kernel/sched/core.c b/aosp/kernel/sched/core.c
index 31fabf8aa5e8..651fa46ac9f5 100644
--- a/aosp/kernel/sched/core.c
+++ b/aosp/aosp/kernel/sched/core.c
@@ -3982,11 +3982,6 @@ bool cpus_share_cache(int this_cpu, int that_cpu)
 
 static inline bool ttwu_queue_cond(struct task_struct *p, int cpu)
 {
-#ifdef CONFIG_SMP
-	if (p->sched_class == &stop_sched_class)
-		return false;
-#endif
-
 	/*
 	 * Do not complicate things with the async wake_list while the CPU is
 	 * in hotplug state.
diff --git a/aosp/kernel/stop_machine.c b/aosp/kernel/stop_machine.c
index 11def33bd5b3..feffcfdc3075 100644
--- a/aosp/kernel/stop_machine.c
+++ b/aosp/aosp/kernel/stop_machine.c
@@ -82,15 +82,18 @@ static void cpu_stop_signal_done(struct cpu_stop_done *done)
 }
 
 static void __cpu_stop_queue_work(struct cpu_stopper *stopper,
-				  struct cpu_stop_work *work)
+					struct cpu_stop_work *work,
+					struct wake_q_head *wakeq)
 {
 	list_add_tail(&work->list, &stopper->works);
+	wake_q_add(wakeq, stopper->thread);
 }
 
 /* queue @work to @stopper.  if offline, @work is completed immediately */
 static bool cpu_stop_queue_work(unsigned int cpu, struct cpu_stop_work *work)
 {
 	struct cpu_stopper *stopper = &per_cpu(cpu_stopper, cpu);
+	DEFINE_WAKE_Q(wakeq);
 	unsigned long flags;
 	bool enabled;
 
@@ -98,13 +101,12 @@ static bool cpu_stop_queue_work(unsigned int cpu, struct cpu_stop_work *work)
 	raw_spin_lock_irqsave(&stopper->lock, flags);
 	enabled = stopper->enabled;
 	if (enabled)
-		__cpu_stop_queue_work(stopper, work);
+		__cpu_stop_queue_work(stopper, work, &wakeq);
 	else if (work->done)
 		cpu_stop_signal_done(work->done);
 	raw_spin_unlock_irqrestore(&stopper->lock, flags);
 
-	if (enabled)
-		wake_up_process(stopper->thread);
+	wake_up_q(&wakeq);
 	preempt_enable();
 
 	return enabled;
@@ -262,6 +264,7 @@ static int cpu_stop_queue_two_works(int cpu1, struct cpu_stop_work *work1,
 {
 	struct cpu_stopper *stopper1 = per_cpu_ptr(&cpu_stopper, cpu1);
 	struct cpu_stopper *stopper2 = per_cpu_ptr(&cpu_stopper, cpu2);
+	DEFINE_WAKE_Q(wakeq);
 	int err;
 
 retry:
@@ -297,8 +300,8 @@ static int cpu_stop_queue_two_works(int cpu1, struct cpu_stop_work *work1,
 	}
 
 	err = 0;
-	__cpu_stop_queue_work(stopper1, work1);
-	__cpu_stop_queue_work(stopper2, work2);
+	__cpu_stop_queue_work(stopper1, work1, &wakeq);
+	__cpu_stop_queue_work(stopper2, work2, &wakeq);
 
 unlock:
 	raw_spin_unlock(&stopper2->lock);
@@ -313,10 +316,7 @@ static int cpu_stop_queue_two_works(int cpu1, struct cpu_stop_work *work1,
 		goto retry;
 	}
 
-	if (!err) {
-		wake_up_process(stopper1->thread);
-		wake_up_process(stopper2->thread);
-	}
+	wake_up_q(&wakeq);
 	preempt_enable();
 
 	return err;
diff --git a/aosp/mm/Kconfig b/aosp/mm/Kconfig
index 3e6b56b17bbb..29856bf29a0f 100644
--- a/aosp/mm/Kconfig
+++ b/aosp/aosp/mm/Kconfig
@@ -1186,6 +1186,17 @@ config IO_MAPPING
 config ARCH_HAS_IOREMAP_PHYS_HOOKS
 	bool
 
+config MEMFD_ASHMEM_SHIM
+	bool "Memfd ashmem ioctl compatibility support"
+	depends on MEMFD_CREATE
+	help
+	  This provides compatibility support for ashmem ioctl commands against
+	  memfd file descriptors. This is useful for compatibility on Android
+	  for older applications that may use ashmem's ioctl commands on the
+	  now memfds passed to them.
+
+	  Unless you are running Android, say N.
+
 config SECRETMEM
 	def_bool ARCH_HAS_SET_DIRECT_MAP && !EMBEDDED
 
diff --git a/aosp/mm/Makefile b/aosp/mm/Makefile
index 7912f6b58bb0..31ec4f350c74 100644
--- a/aosp/mm/Makefile
+++ b/aosp/aosp/mm/Makefile
@@ -134,6 +134,7 @@ obj-$(CONFIG_PERCPU_STATS) += percpu-stats.o
 obj-$(CONFIG_ZONE_DEVICE) += memremap.o
 obj-$(CONFIG_HMM_MIRROR) += hmm.o
 obj-$(CONFIG_MEMFD_CREATE) += memfd.o
+obj-$(CONFIG_MEMFD_ASHMEM_SHIM) += memfd-ashmem-shim.o
 obj-$(CONFIG_MAPPING_DIRTY_HELPERS) += mapping_dirty_helpers.o
 obj-$(CONFIG_PTDUMP_CORE) += ptdump.o
 obj-$(CONFIG_PAGE_REPORTING) += page_reporting.o
diff --git a/aosp/mm/filemap.c b/aosp/mm/filemap.c
index 1a0e07bbab2e..1d09853c5473 100644
--- a/aosp/mm/filemap.c
+++ b/aosp/aosp/mm/filemap.c
@@ -2042,7 +2042,6 @@ struct folio *__filemap_get_folio(struct address_space *mapping, pgoff_t index,
 			folio_unlock(folio);
 	}
 
-	trace_android_vh_filemap_get_folio_end(mapping, folio);
 	return folio;
 }
 EXPORT_SYMBOL(__filemap_get_folio);
diff --git a/aosp/mm/memfd-ashmem-shim-internal.h b/aosp/mm/memfd-ashmem-shim-internal.h
new file mode 100644
index 000000000000..b499434a94c7
--- /dev/null
+++ b/aosp/aosp/mm/memfd-ashmem-shim-internal.h
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * Ashmem compatability for memfd
+ *
+ * Copyright (c) 2025, Google LLC.
+ * Author: Isaac J. Manjarres <isaacmanjarres@google.com>
+ */
+
+#ifndef _MM_MEMFD_ASHMEM_SHIM_INTERNAL_H
+#define _MM_MEMFD_ASHMEM_SHIM_INTERNAL_H
+
+#include <linux/compat.h>
+#include <linux/ioctl.h>
+#include <linux/types.h>
+
+#define ASHMEM_NAME_LEN		256
+
+/* Return values from ASHMEM_PIN: Was the mapping purged while unpinned? */
+#define ASHMEM_NOT_PURGED	0
+#define ASHMEM_WAS_PURGED	1
+
+/* Return values from ASHMEM_GET_PIN_STATUS: Is the mapping pinned? */
+#define ASHMEM_IS_UNPINNED	0
+#define ASHMEM_IS_PINNED	1
+
+struct ashmem_pin {
+	__u32 offset;	/* offset into region, in bytes, page-aligned */
+	__u32 len;	/* length forward from offset, in bytes, page-aligned */
+};
+
+#define __ASHMEMIOC		0x77
+
+#define ASHMEM_SET_NAME		_IOW(__ASHMEMIOC, 1, char[ASHMEM_NAME_LEN])
+#define ASHMEM_GET_NAME		_IOR(__ASHMEMIOC, 2, char[ASHMEM_NAME_LEN])
+#define ASHMEM_SET_SIZE		_IOW(__ASHMEMIOC, 3, size_t)
+#define ASHMEM_GET_SIZE		_IO(__ASHMEMIOC, 4)
+#define ASHMEM_SET_PROT_MASK	_IOW(__ASHMEMIOC, 5, unsigned long)
+#define ASHMEM_GET_PROT_MASK	_IO(__ASHMEMIOC, 6)
+#define ASHMEM_PIN		_IOW(__ASHMEMIOC, 7, struct ashmem_pin)
+#define ASHMEM_UNPIN		_IOW(__ASHMEMIOC, 8, struct ashmem_pin)
+#define ASHMEM_GET_PIN_STATUS	_IO(__ASHMEMIOC, 9)
+#define ASHMEM_PURGE_ALL_CACHES	_IO(__ASHMEMIOC, 10)
+#define ASHMEM_GET_FILE_ID		_IOR(__ASHMEMIOC, 11, unsigned long)
+
+/* support of 32bit userspace on 64bit platforms */
+#ifdef CONFIG_COMPAT
+#define COMPAT_ASHMEM_SET_SIZE		_IOW(__ASHMEMIOC, 3, compat_size_t)
+#define COMPAT_ASHMEM_SET_PROT_MASK	_IOW(__ASHMEMIOC, 5, unsigned int)
+#endif
+
+#endif /* _MM_MEMFD_ASHMEM_SHIM_INTERNAL_H */
diff --git a/aosp/mm/memfd-ashmem-shim.c b/aosp/mm/memfd-ashmem-shim.c
new file mode 100644
index 000000000000..258498cca9bb
--- /dev/null
+++ b/aosp/aosp/mm/memfd-ashmem-shim.c
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * Ashmem compatability for memfd
+ *
+ * Copyright (c) 2025, Google LLC.
+ * Author: Isaac J. Manjarres <isaacmanjarres@google.com>
+ */
+
+#include <asm-generic/mman-common.h>
+#include <linux/capability.h>
+#include <linux/fs.h>
+#include <linux/memfd.h>
+#include <linux/uaccess.h>
+
+#include "memfd-ashmem-shim.h"
+#include "memfd-ashmem-shim-internal.h"
+
+/* memfd file names all start with memfd: */
+#define MEMFD_PREFIX "memfd:"
+#define MEMFD_PREFIX_LEN (sizeof(MEMFD_PREFIX) - 1)
+
+static const char *get_memfd_name(struct file *file)
+{
+	/* This pointer is always valid, so no need to check if it's NULL. */
+	const char *file_name = file->f_path.dentry->d_name.name;
+
+	if (file_name != strstr(file_name, MEMFD_PREFIX))
+		return NULL;
+
+	return file_name;
+}
+
+static long get_name(struct file *file, void __user *name)
+{
+	const char *file_name = get_memfd_name(file);
+	size_t len;
+
+	if (!file_name)
+		return -EINVAL;
+
+	/* Strip MEMFD_PREFIX to retain compatibility with ashmem driver. */
+	file_name = &file_name[MEMFD_PREFIX_LEN];
+
+	/*
+	 * The expectation is that the user provided buffer is ASHMEM_NAME_LEN in size, which is
+	 * larger than the maximum size of a name for a memfd buffer, so the name should always fit
+	 * within the given buffer.
+	 *
+	 * However, we should ensure that the string will indeed fit in the user provided buffer.
+	 *
+	 * Add 1 to the copy size to account for the NUL terminator
+	 */
+	len = strlen(file_name) + 1;
+	if (len > ASHMEM_NAME_LEN)
+		return -EINVAL;
+
+	return copy_to_user(name, file_name, len) ? -EFAULT : 0;
+}
+
+static long get_prot_mask(struct file *file)
+{
+	long prot_mask = PROT_READ | PROT_EXEC;
+	long seals = memfd_fcntl(file, F_GET_SEALS, 0);
+
+	if (seals < 0)
+		return seals;
+
+	/* memfds are readable and executable by default. Only writability can be changed. */
+	if (!(seals & (F_SEAL_WRITE | F_SEAL_FUTURE_WRITE)))
+		prot_mask |= PROT_WRITE;
+
+	return prot_mask;
+}
+
+static long set_prot_mask(struct file *file, unsigned long prot)
+{
+	long curr_prot = get_prot_mask(file);
+	long ret = 0;
+
+	if (curr_prot < 0)
+		return curr_prot;
+
+	/*
+	 * memfds are always readable and executable; there is no way to remove either mapping
+	 * permission, nor is there a known usecase that requires it.
+	 *
+	 * Attempting to remove either of these mapping permissions will return successfully, but
+	 * will be a nop, as the buffer will still be mappable with these permissions.
+	 */
+	prot |= PROT_READ | PROT_EXEC;
+
+	/* Only allow permissions to be removed. */
+	if ((curr_prot & prot) != prot)
+		return -EINVAL;
+
+	/*
+	 * Removing PROT_WRITE:
+	 *
+	 * We could prevent any other mappings from having write permissions by adding the
+	 * F_SEAL_WRITE mapping. However, that would conflict with known usecases where it is
+	 * desirable to maintain an existing writable mapping, but forbid future writable mappings.
+	 *
+	 * To support those usecases, we use F_SEAL_FUTURE_WRITE.
+	 */
+	if (!(prot & PROT_WRITE))
+		ret = memfd_fcntl(file, F_ADD_SEALS, F_SEAL_FUTURE_WRITE);
+
+	return ret;
+}
+
+/*
+ * memfd_ashmem_shim_ioctl - ioctl handler for ashmem commands
+ * @file: The shmem file.
+ * @cmd: The ioctl command.
+ * @arg: The argument for the ioctl command.
+ *
+ * The purpose of this handler is to allow old applications to continue working
+ * on newer kernels by allowing them to invoke ashmem ioctl commands on memfds.
+ *
+ * The ioctl handler attempts to retain as much compatibility with the ashmem
+ * driver as possible.
+ */
+long memfd_ashmem_shim_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	long ret = -ENOTTY;
+	unsigned long inode_nr;
+
+	switch (cmd) {
+	/*
+	 * Older applications won't create memfds and try to use ASHMEM_SET_NAME/ASHMEM_SET_SIZE on
+	 * them intentionally.
+	 *
+	 * Instead, we can end up in this scenario if an old application receives a memfd that was
+	 * created by another process.
+	 *
+	 * However, the current process shouldn't expect to be able to reliably [re]name/size a
+	 * buffer that was shared with it, since the process that shared that buffer with it, or
+	 * any other process that references the buffer could have already mapped it.
+	 *
+	 * Additionally in the case of ASHMEM_SET_SIZE, when processes create memfds that are going
+	 * to be shared with other processes in Android, they also specify the size of the memory
+	 * region and seal the file against any size changes. Therefore, ASHMEM_SET_SIZE should not
+	 * be supported anyway.
+	 *
+	 * Therefore, it is reasonable to return -EINVAL here, as if the buffer was already mapped.
+	 */
+	case ASHMEM_SET_NAME:
+	case ASHMEM_SET_SIZE:
+		ret = -EINVAL;
+		break;
+	case ASHMEM_GET_NAME:
+		ret = get_name(file, (void __user *)arg);
+		break;
+	case ASHMEM_GET_SIZE:
+		ret = i_size_read(file_inode(file));
+		break;
+	case ASHMEM_SET_PROT_MASK:
+		ret = set_prot_mask(file, arg);
+		break;
+	case ASHMEM_GET_PROT_MASK:
+		ret = get_prot_mask(file);
+		break;
+	/*
+	 * Unpinning ashmem buffers was deprecated with the release of Android 10,
+	 * as it did not yield any remarkable benefits. Therefore, ignore pinning
+	 * related requests.
+	 *
+	 * This makes it so that memory is always "pinned" or never entirely freed
+	 * until all references to the ashmem buffer are dropped. The memory occupied
+	 * by the buffer is still subject to being reclaimed (swapped out) under memory
+	 * pressure, but that is not the same as being freed.
+	 *
+	 * This makes it so that:
+	 *
+	 * 1. Memory is always pinned and therefore never purged.
+	 * 2. Requests to unpin memory (make it a candidate for being freed) are ignored.
+	 */
+	case ASHMEM_PIN:
+		ret = ASHMEM_NOT_PURGED;
+		break;
+	case ASHMEM_UNPIN:
+		ret = 0;
+		break;
+	case ASHMEM_GET_PIN_STATUS:
+		ret = ASHMEM_IS_PINNED;
+		break;
+	case ASHMEM_PURGE_ALL_CACHES:
+		ret = capable(CAP_SYS_ADMIN) ? 0 : -EPERM;
+		break;
+	case ASHMEM_GET_FILE_ID:
+		inode_nr = file_inode(file)->i_ino;
+		if (copy_to_user((void __user *)arg, &inode_nr, sizeof(inode_nr)))
+			ret = -EFAULT;
+		else
+			ret = 0;
+		break;
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_COMPAT
+long memfd_ashmem_shim_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	if (cmd == COMPAT_ASHMEM_SET_SIZE)
+		cmd = ASHMEM_SET_SIZE;
+	else if (cmd == COMPAT_ASHMEM_SET_PROT_MASK)
+		cmd = ASHMEM_SET_PROT_MASK;
+
+	return memfd_ashmem_shim_ioctl(file, cmd, arg);
+}
+#endif
diff --git a/aosp/mm/memfd-ashmem-shim.h b/aosp/mm/memfd-ashmem-shim.h
new file mode 100644
index 000000000000..026789b0344b
--- /dev/null
+++ b/aosp/aosp/mm/memfd-ashmem-shim.h
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __MM_MEMFD_ASHMEM_SHIM_H
+#define __MM_MEMFD_ASHMEM_SHIM_H
+
+/*
+ * mm/memfd-ashmem-shim.h
+ *
+ * Ashmem compatability for memfd
+ *
+ * Copyright (c) 2025, Google LLC.
+ * Author: Isaac J. Manjarres <isaacmanjarres@google.com>
+ *
+ */
+
+#include <linux/fs.h>
+
+long memfd_ashmem_shim_ioctl(struct file *file, unsigned int cmd, unsigned long arg);
+#ifdef CONFIG_COMPAT
+long memfd_ashmem_shim_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg);
+#endif
+#endif /* __MM_MEMFD_ASHMEM_SHIM_H */
diff --git a/aosp/mm/memory.c b/aosp/mm/memory.c
index 043cdd950125..78216a14ff5d 100644
--- a/aosp/mm/memory.c
+++ b/aosp/aosp/mm/memory.c
@@ -169,8 +169,6 @@ static int __init init_zero_pfn(void)
 }
 early_initcall(init_zero_pfn);
 
-EXPORT_TRACEPOINT_SYMBOL_GPL(rss_stat);
-
 void mm_trace_rss_stat(struct mm_struct *mm, int member, long count)
 {
 	trace_rss_stat(mm, member, count);
diff --git a/aosp/mm/page_alloc.c b/aosp/mm/page_alloc.c
index 08c892e34e1a..11dd008e254b 100644
--- a/aosp/mm/page_alloc.c
+++ b/aosp/aosp/mm/page_alloc.c
@@ -5342,7 +5342,7 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 	const bool costly_order = order > PAGE_ALLOC_COSTLY_ORDER;
 	struct page *page = NULL;
 	unsigned int alloc_flags;
-	unsigned long did_some_progress = 0;
+	unsigned long did_some_progress;
 	enum compact_priority compact_priority;
 	enum compact_result compact_result;
 	int compaction_retries;
@@ -5352,10 +5352,6 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 	int reserve_flags;
 	unsigned long alloc_start = jiffies;
 	bool should_alloc_retry = false;
-	unsigned long pages_reclaimed = 0;
-	int retry_loop_count = 0;
-	u64 stime = 0;
-
 	/*
 	 * We also sanity check to catch abuse of atomic reserves being used by
 	 * callers that are not in atomic context.
@@ -5364,8 +5360,6 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 				(__GFP_ATOMIC|__GFP_DIRECT_RECLAIM)))
 		gfp_mask &= ~__GFP_ATOMIC;
 
-	trace_android_vh_alloc_pages_slowpath_start(&stime);
-
 restart:
 	compaction_retries = 0;
 	no_progress_loops = 0;
@@ -5486,7 +5480,6 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 	}
 
 retry:
-	retry_loop_count++;
 	/*
 	 * Deal with possible cpuset update races or zonelist updates to avoid
 	 * infinite retries.
@@ -5542,7 +5535,6 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 	/* Try direct reclaim and then allocating */
 	page = __alloc_pages_direct_reclaim(gfp_mask, order, alloc_flags, ac,
 							&did_some_progress);
-	pages_reclaimed += did_some_progress;
 	if (page)
 		goto got_pg;
 
@@ -5666,8 +5658,6 @@ __alloc_pages_slowpath(gfp_t gfp_mask, unsigned int order,
 			"page allocation failure: order:%u", order);
 got_pg:
 	trace_android_vh_alloc_pages_slowpath(gfp_mask, order, alloc_start);
-	trace_android_vh_alloc_pages_slowpath_end(&gfp_mask, order, alloc_start,
-			stime, did_some_progress, pages_reclaimed, retry_loop_count);
 	return page;
 }
 
@@ -5892,7 +5882,6 @@ struct page *__alloc_pages(gfp_t gfp, unsigned int order, int preferred_nid,
 	unsigned int alloc_flags = ALLOC_WMARK_LOW;
 	gfp_t alloc_gfp; /* The gfp_t that was actually used for allocation */
 	struct alloc_context ac = { };
-	u64 stime = 0;
 
 	/*
 	 * There are several places where we assume that the order value is sane
@@ -5901,8 +5890,6 @@ struct page *__alloc_pages(gfp_t gfp, unsigned int order, int preferred_nid,
 	if (WARN_ON_ONCE_GFP(order >= MAX_ORDER, gfp))
 		return NULL;
 
-	trace_android_vh_alloc_pages_start(&stime);
-
 	gfp &= gfp_allowed_mask;
 	/*
 	 * Apply scoped allocation constraints. This is mainly about GFP_NOFS
@@ -5951,7 +5938,6 @@ struct page *__alloc_pages(gfp_t gfp, unsigned int order, int preferred_nid,
 
 	trace_mm_page_alloc(page, order, alloc_gfp, ac.migratetype);
 	kmsan_alloc_page(page, order, alloc_gfp);
-	trace_android_vh_alloc_pages_end(order, alloc_gfp, stime);
 
 	return page;
 }
diff --git a/aosp/mm/shmem.c b/aosp/mm/shmem.c
index 762354fb17e4..2b6cef108b2c 100644
--- a/aosp/mm/shmem.c
+++ b/aosp/aosp/mm/shmem.c
@@ -89,6 +89,10 @@ static struct vfsmount *shm_mnt;
 
 #include "internal.h"
 
+#ifdef CONFIG_MEMFD_ASHMEM_SHIM
+#include "memfd-ashmem-shim.h"
+#endif
+
 #define BLOCKS_PER_PAGE  (PAGE_SIZE/512)
 #define VM_ACCT(size)    (PAGE_ALIGN(size) >> PAGE_SHIFT)
 
@@ -3976,6 +3980,12 @@ static const struct file_operations shmem_file_operations = {
 	.splice_write	= iter_file_splice_write,
 	.fallocate	= shmem_fallocate,
 #endif
+#ifdef CONFIG_MEMFD_ASHMEM_SHIM
+	.unlocked_ioctl	= memfd_ashmem_shim_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= memfd_ashmem_shim_compat_ioctl,
+#endif
+#endif
 };
 
 static const struct inode_operations shmem_inode_operations = {
diff --git a/aosp/mm/vmscan.c b/aosp/mm/vmscan.c
index ac95060f4991..7e3d68c4e5d2 100644
--- a/aosp/mm/vmscan.c
+++ b/aosp/aosp/mm/vmscan.c
@@ -1343,7 +1343,6 @@ static int __remove_mapping(struct address_space *mapping, struct folio *folio,
 	BUG_ON(!folio_test_locked(folio));
 	BUG_ON(mapping != folio_mapping(folio));
 
-	trace_android_vh_remove_mapping(mapping, folio, reclaimed);
 	if (!folio_test_swapcache(folio))
 		spin_lock(&mapping->host->i_lock);
 	xa_lock_irq(&mapping->i_pages);
@@ -1430,7 +1429,6 @@ static int __remove_mapping(struct address_space *mapping, struct folio *folio,
 	xa_unlock_irq(&mapping->i_pages);
 	if (!folio_test_swapcache(folio))
 		spin_unlock(&mapping->host->i_lock);
-	trace_android_vh_remove_mapping_failed(mapping, folio, reclaimed);
 	return 0;
 }
 


